programs/perpetuals/src/state/custody.rs

```diff
@@ -181,6 +211,62 @@ impl Custody {
             && self.oracle.validate()
             && self.pricing.validate()
             && self.fees.validate()
+            && self.borrow_rate.validate()
+    }
+
+    pub fn update_borrow_rate(&mut self, curtime: i64) -> Result<()> {
+        // if current_utilization < optimal_utilization:
+        //   rate = base_rate + (current_utilization / optimal_utilization) * slope1
+        // else:
+        //   rate = base_rate + slope1 + (current_utilization - optimal_utilization) / (1 - optimal_utilization) * slope2
+
+        if curtime <= self.borrow_rate_state.last_update {
+            return Ok(());
+        }
+
+        let current_utilization = math::checked_div(
+            math::checked_mul(self.assets.locked as u128, Perpetuals::RATE_POWER)?,
+            self.assets.owned as u128,
+        )?;
+
+        let hourly_rate = if current_utilization < (self.borrow_rate.optimal_utilization as u128)
+            || (self.borrow_rate.optimal_utilization as u128) >= Perpetuals::RATE_POWER
+        {
+            math::checked_div(
+                math::checked_mul(current_utilization, self.borrow_rate.slope1 as u128)?,
```

Comment:
would it better to divide before mult just in case the multiplication overflows? Will most likely don't but you never know

Reply:
not really, if you divide first you lose the precision, because this is an integer division. 
for example:
12345 / 10000 * 10000 = 10000
but
12345 * 10000 / 10000 = 12345
to ensure multiplication doesn't overflow I cast to u128 first

Reply:
ah yes because utilization and slope are max u64::MAX so the product will be less than U128::MAX.